### PR TITLE
[Feature:Autograding] Assignment Configuration Schema

### PR DESCRIPTION
**Pull request by emaicus**
_Edited 7/6/19_

### This PR:
1. Adds  ```complete_config_schema.json```, an initial json schema for complete configuration files for assignments.
    1. This schema casts a wide net, and allows some errors to slip through. This can be tightened/iterated on.
    2. This schema is housed in a new directory ```{SUBMITTY_INSTALL_DIR}/bin/json_schemas```
2. Adds a new module to ```python_submitty_utils```, the  ```submitty_schema_validator```. This module can be iterated on/extended as new shemas are added to the system.
3.  Adds a new testcase to all relevant ```C++``` integration tests, which checks that the complete config used by the test conforms to  the ```complete_config_schema.json```.
4. Adds a migration to add installs of the pip packages ```jsonschema``` and ```jsonref``` to the system. Also installs these dependencies in travis and ```install_system.sh```
    1. This migration may need editing if issue #4011 is addressed first. 
5. Addresses a handful of errors detected by the new schema testcases.

### CODE REVIEW

File: json_schemas/sample_unit_tests.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      config_json = json.load(infile)
+
+    try:
+      submitty_json_validator.submitty_validate_schemas(config_json,SCHEMA)
+      success = True
+    except Exception as e:
+      print(e)
+      success = False
+
+    self.assertTrue(success)
```
Comment (MasterOdin) [inline json_schemas/sample_unit_tests.py]: You can write this much more succinctly using [unittest.subTest](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest) (I eyeballed this code so may need some slight tweaking):
```
import unittest
import schema_validator as submitty_json_validator
import json
import os 

test_files = [
    'complete_config_01_simple_python',
    'complete_config_02_simple_cpp',
    'complete_config_03_multipart',
    'complete_config_04_python_static_analysis',
    'complete_config_05_cpp_static_analysis',
    'complete_config_06_loop_types',
    'complete_config_07_loop_depth,
    'complete_config_08_memory_debugging',
    'complete_config_09_java_testing',
    'complete_config_10_java_coverage',
    'complete_config_11_resources',
    'complete_config_12_system_calls',
    'complete_config_13_cmake_compilation'
]

class TestSchema(unittest.Testcase):
    def test_validate_schema(self):
        for test_file in test_files:
            with self.subTest(msg='Testing validation of test_file', test_file=test_file):
                with open(os.path.join('tutorial_configs', test_file), 'r') as infile:
                    config_json = json.load(infile)

                try:
                    submitty_json_validator.submitty_validate_schemas(config_json, SCHEMA)
                    success = True
                except Exception as e:
                    print(e)
                    success = False
                self.assertTrue(success)
```

Comment (emaicus) [reply json_schemas/sample_unit_tests.py]: Thanks very much for the tip! I figured there must be a less verbose way to write these when there is so much overlap. I'll get the change in soon.

File: json_schemas/schema_validator.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  try:
+    #validate.
+    validate(instance=my_json_chunk, schema=my_schema)
+  except Exception as e:
+    #If the item was required, print an error. Otherwise, the exception will be suppressed later, so don't be loud. 
+    if required:
+      traceback.print_exc()
+    if descriptive_title != '':
+      print('ERROR: {0} was not properly formatted.'.format(descriptive_title))
+    raise Exception(e.message)
```
Comment (MasterOdin) [inline json_schemas/schema_validator.py]: you can change this to be just `raise` and python will re-raise whatever exception put us here.

File: json_schemas/schema_validator.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      submitty_validate_schema(validator, abstract_validation_object_schema, None, None, descriptive_title='testcase {0} validator {1} (generic)'.format(testcase_num, validator_num))
+      submitty_validate_schema(validator, specific_validation_object_schema, None, None, descriptive_title='testcase {0} validator {1} (specific)'.format(testcase_num, validator_num))
+      validator_num += 1
+
+    testcase_num+=1
+  #Finally, validate the config as a whole.
+  submitty_validate_schema(config_json, schema, None, None, descriptive_title='whole schema')
+
+#We validate a key within a config and a schema. If no key is given, we evaluate the whole config or schema.
+def submitty_validate_schema(config_json, schema, config_json_key, schema_key, descriptive_title='', required=True):
```
Comment (MasterOdin) [inline json_schemas/schema_validator.py]: I'd add a default value of `None` for `config_json_key` and `schema_key` here. Then above, anytime you would have passed in `None` to the function, you can just drop the arguments (e.g. on line 68 and 69).

File: json_schemas/schema_validator.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  parser.add_argument('--config', required=True, help='The path to the configuration to be validated')
+  parser.add_argument('--schema', required=True, help='The path to the schema to be validated against')
+  return parser.parse_args()
+
+
+'''
+Validates a config json piece by piece.
+'''
+def submitty_validate_schemas(config_json, schema):
+  submitty_validate_schema(config_json, schema['properties'], 'autograding', 'autograding', descriptive_title='global autograding block')
```
Comment (MasterOdin) [inline json_schemas/schema_validator.py]: You don't need to give a named argument here as it's the next one to be used anyway. I'd suggest removing the `descriptive_title=` bit here (and other places where its preceded by 4 arguments) and save the characters.

**Comment by emaicus** [comment]
@MasterOdin @bmcutler, This PR is ready for re-review, and the initial comment has been updated to reflect its current state.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                                      ('ERROR: {0} was not properly '
+                                       'formatted')
+                                      .format(descriptive_title),
+                                      descriptive_title,
+                                      e)
+
+
+def validate_complete_config_schema_using_filenames(config_path,
+                                                    schema_path,
+                                                    show_warnings=True):
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Where is this going to be used where `show_warnings` is going to be True? It looks like all current usage it is set to `False`?

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+    def print_human_readable_error(self):
+        """Use to print a human readable version of this exception."""
+        print()
+        print('{0}:'.format(self.my_message))
+        if self.schema_message is not None:
+            print(self.schema_message)
+            print(("The portion of your configuration that caused "
+                   "the error is:"))
+            print(json.dumps(self.config_json, indent=4))
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: This should probably be printed to stderr

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+def complete_config_validator(j_, s_, show_warnings=True):
+    """
+    Validate a complete configuration against a schema.
+
+    Given a complete config and a complete config schema, validate
+    the complete config one piece at a time. On error, raise a
+    SubmittySchemaException, which is capable of printing a
+    human readable error message.
+    """
+    warn = show_warnings
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Just name the parameter warn

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            validate_schema(t, p_s, 'title', t_name, warn=warn)
+            validate_schema(t, p_s, 'hidden', t_name, warn=warn)
+            validate_schema(t, p_s, 'extra_credit', t_name, warn=warn)
+            validate_schema(t, p_s, 'input_generation_commands', t_name,
+                            warn=warn)
+            validate_schema(t, p_s, 'executable_name', t_name, warn=warn)
+            validate_schema(t, p_s, 'testcase_label', t_name, warn=warn)
+
+        except Exception:
+            raise
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: same as above

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        containers = t['containers']
+        # validate each container in the testcase
+        for c in containers:
+            try:
+                validate_schema(c, c_schema, prefix=t_name, warn=warn)
+            except SubmittySchemaException:
+                raise
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: same here

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+        validators = t['validation']
+        validator_num = 0
+        # Validate each validation object in the testcase
+        for v in validators:
+            try:
+                validate_schema(v, abs_v_schema, prefix=t_name, warn=warn)
+                validate_schema(v, spec_v_schema, prefix=t_name, warn=warn)
+            except SubmittySchemaException:
+                raise
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: same here

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            validator_num += 1
+        testcase_num += 1
+
+        try:
+            validate_schema(t, t_s, prefix=t_name, warn=warn)
+        except SubmittySchemaException:
+            raise
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: same here

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    # Finally, validate the config as a whole.
+    try:
+        validate_schema(j_, s_, prefix='Your config json', warn=warn)
+    except SubmittySchemaException:
+        raise
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: same here

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+                                      None)
+    if my_json_chunk is None:
+        if warn:
+            print(("WARNING: could not identify "
+                   "{0} ({1})").format(descriptive_title, key))
+        return
+    try:
+        # validate.
+        validate(instance=my_json_chunk, schema=my_schema)
+    except Exception as e:
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: make this less broad. what are the exceptions validate can throw?

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+    else:
+        my_json_chunk = j_
+        my_schema = s_
+    # If attempting to validate an item not specified in the schema, fail.
+    if my_schema is None:
+        raise SubmittySchemaException(j_,
+                                      s_,
+                                      ("ERROR: There is no specification for "
+                                       "{0} in the schema. Please "
+                                       "add a specification.").format(key),
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: set the message in a variable above this line to make it a bit more readable and potentially edit.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+        return
+    try:
+        # validate.
+        validate(instance=my_json_chunk, schema=my_schema)
+    except Exception as e:
+        raise SubmittySchemaException(my_json_chunk,
+                                      my_schema,
+                                      ('ERROR: {0} was not properly '
+                                       'formatted')
+                                      .format(descriptive_title),
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: same as above

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+    Given a path to a complete configuration and a schema, this function
+    loads both and validates the configuration against the schema. On
+    failure, an exception is thrown, else, nothing is returned.
+    """
+    with open(schema_path, 'r') as infile:
+        schema = jsonref.load(infile)
+
+    with open(config_path, 'r') as infile:
+        config_json = jsonref.load(infile)
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Are we allowing the config jsons to have references or must they be standard json?

File: tests/integrationTests/tests/c_failure_messages/__init__.py [hunk 17]
```diff
@@ -5,12 +5,13 @@
...
 import shutil
+from submitty_utils import submitty_schema_validator
 
 ############################################################################
 # COPY THE ASSIGNMENT FROM THE SAMPLE ASSIGNMENTS DIRECTORIES
 
-SAMPLE_ASSIGNMENT_CONFIG = SUBMITTY_INSTALL_DIR + "/more_autograding_examples/c_failure_messages/config"
-SAMPLE_SUBMISSIONS       = SUBMITTY_INSTALL_DIR + "/more_autograding_examples/c_failure_messages/submissions"
+SAMPLE_ASSIGNMENT_CONFIG = os.path.join(SUBMITTY_INSTALL_DIR, "more_autograding_examples/c_failure_messages/config")
+SAMPLE_SUBMISSIONS       = os.path.join(SUBMITTY_INSTALL_DIR, "more_autograding_examples/c_failure_messages/submissions")
```
Comment (MasterOdin) [inline tests/integrationTests/tests/c_failure_messages/__init__.py]: If we're going to fix this to use join, let's separate out each part into its own string and passed in and do it right.

File: tests/integrationTests/tests/python_linehighlight/__init__.py [hunk 18]
```diff
@@ -49,6 +50,15 @@ def initialize(test):
...
 
+@testcase
+def schema_validation(test):
+    config_path = os.path.join(test.testcase_path, 'assignment_config', 'complete_config.json')
+    schema = os.path.join(SUBMITTY_INSTALL_DIR, 'bin', 'json_schemas', 'complete_config_schema.json')
+    try:
+        submitty_schema_validator.validate_complete_config_schema_using_filenames(config_path, schema, show_warnings=False)
+    except submitty_schema_validator.SubmittySchemaException as s:
+        s.print_human_readable_error()
+        raise
```
Comment (MasterOdin) [inline tests/integrationTests/tests/python_linehighlight/__init__.py]: extract this block of code into a function in `lib.py` and just call that with just a parameter of the config_path instead of having this just be copied and pasted in each.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    warn = show_warnings
+    s_prop = s_['properties']
+    try:
+        validate_schema(j_, s_prop, 'autograding', 'global', warn=warn)
+        validate_schema(j_, s_prop, 'autograding_method', 'global', warn=warn)
+        validate_schema(j_, s_prop, 'container_options', 'global', warn=warn)
+        validate_schema(j_, s_prop, 'resource_limits', 'global', warn=warn)
+    except SubmittySchemaException:
+        raise
```
Comment (MasterOdin) [inline python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Get rid of the try/catch if you're just immediately re-raising

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+            validate_schema(t, p_s, 'single_port_per_container', t_name,
+                            warn=warn)
+            validate_schema(t, p_s, 'use_router', t_name, warn=warn)
+            validate_schema(t, p_s, 'title', t_name, warn=warn)
+            validate_schema(t, p_s, 'hidden', t_name, warn=warn)
+            validate_schema(t, p_s, 'extra_credit', t_name, warn=warn)
+            validate_schema(t, p_s, 'input_generation_commands', t_name,
+                            warn=warn)
+            validate_schema(t, p_s, 'executable_name', t_name, warn=warn)
+            validate_schema(t, p_s, 'testcase_label', t_name, warn=warn)
```
[the remark this answers is not in the document]
Comment (emaicus) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Fixed in upcoming commit.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py (hunk 6 above)
Comment (emaicus) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Fixed in upcoming commit.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py (hunk 7 above)
Comment (emaicus) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Fixed in upcoming commit.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py (hunk 14 above)
Comment (emaicus) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Fixed in upcoming commit.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py (hunk 16 above)
Comment (emaicus) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Good catch. At the moment, configs must be standard JSON. Schemas can have references, however, as that makes them far more writable.

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py (hunk 5 above)
Comment (emaicus) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: Changed the default to be ```False```.

File: tests/integrationTests/tests/python_linehighlight/__init__.py (hunk 18 above)
Comment (emaicus) [reply tests/integrationTests/tests/python_linehighlight/__init__.py]: Fixed in upcoming commit.

File: tests/integrationTests/tests/python_linehighlight/__init__.py [hunk 21]
```diff
@@ -49,6 +49,10 @@ def initialize(test):
 
 ############################################################################
 
+@testcase
+def schema_validation(test):
+    config_path = os.path.join(test.testcase_path, 'assignment_config', 'complete_config.json')
```
Comment (MasterOdin) [inline tests/integrationTests/tests/python_linehighlight/__init__.py]: Should this test have `cleanup(test)` at the start of it like the others?

File: tests/integrationTests/tests/tutorial_01_simple_python/__init__.py [hunk 22]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_01_simple_python/__init__.py]: Can revert this change

File: tests/integrationTests/tests/tutorial_02_simple_cpp/__init__.py [hunk 23]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_02_simple_cpp/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_03_multipart/__init__.py [hunk 24]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_03_multipart/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_04_python_static_analysis/__init__.py [hunk 25]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_04_python_static_analysis/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_05_cpp_static_analysis/__init__.py [hunk 26]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_05_cpp_static_analysis/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_06_loop_types/__init__.py [hunk 27]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_06_loop_types/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_07_loop_depth/__init__.py [hunk 28]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_07_loop_depth/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_08_memory_debugging/__init__.py [hunk 29]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_08_memory_debugging/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_09_java_testing/__init__.py [hunk 30]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_09_java_testing/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_11_resources/__init__.py [hunk 31]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_11_resources/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_12_system_calls/__init__.py [hunk 32]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_12_system_calls/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_16_docker_network_python/__init__.py [hunk 33]
```diff
@@ -1,5 +1,5 @@
 # Necessary imports. Provides library functions to ease writing tests.
-from lib import prebuild, testcase, SUBMITTY_TUTORIAL_DIR
+from lib import prebuild, testcase, SUBMITTY_INSTALL_DIR, SUBMITTY_TUTORIAL_DIR
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_16_docker_network_python/__init__.py]: can revert this change

File: tests/integrationTests/tests/tutorial_16_docker_network_python/__init__.py [hunk 34]
```diff
@@ -1,5 +1,5 @@
...
 
 import subprocess
 import os
```
Comment (MasterOdin) [inline tests/integrationTests/tests/tutorial_16_docker_network_python/__init__.py]: Should this test have a schema validation testcase, or is that something you're getting to later?

File: python_submitty_utils/submitty_utils/submitty_schema_validator.py (hunk 16 above)
Comment (MasterOdin) [reply python_submitty_utils/submitty_utils/submitty_schema_validator.py]: I'm very much against supporting jsonref in regular configs as it's not a standard feature and that there's other formats that already support that and are better suited for what we're doing, such as YAML.

File: tests/integrationTests/tests/python_linehighlight/__init__.py (hunk 21 above)
Comment (emaicus) [reply tests/integrationTests/tests/python_linehighlight/__init__.py]: This testcase doesn't have a cleanup function.

Comment (MasterOdin) [reply tests/integrationTests/tests/python_linehighlight/__init__.py]: I think a better question is why does most of the tests then have a cleanup function and this one doesn't? Should it have a cleanup? Should the cleanup be moved to be a pre-processor step for any functions run under the `@testcase` annotation?

File: tests/integrationTests/tests/tutorial_16_docker_network_python/__init__.py (hunk 34 above)
Comment (emaicus) [reply tests/integrationTests/tests/tutorial_16_docker_network_python/__init__.py]: I'll add that one later. PR #4151 will begins to clear a path towards allowing us to have integration tests of docker and docker network grading. When those testcases are added, I'll add the schema validation.
